internal/storage/bucket_handle.go

```diff
@@ -430,6 +435,35 @@ func (b *bucketHandle) ComposeObjects(ctx context.Context, req *gcs.ComposeObjec
 	return
 }
 
+// Fetch bucket type from GetStorageLayout and set it. e.g. Hierarchical or NonHierarchical
+func (b *bucketHandle) FetchAndSetBucketType() {
+	if b.controlClient == nil {
+		b.bucketType = gcs.NonHierarchical
+		return
+	}
+
+	var callOptions []gax.CallOption
+	stoargeLayout, err := b.controlClient.GetStorageLayout(context.Background(), &controlpb.GetStorageLayoutRequest{
+		Name:      "projects/_/buckets/" + b.bucketName + "/storageLayout",
+		Prefix:    "",
+		RequestId: "",
+	}, callOptions...)
+
+	// In case bucket does not exist, set type unknown instead of panic.
+	if err != nil {
+		b.bucketType = gcs.Unknown
+		log.Printf("GetStorageLayout: %v", err)
+		return
+	}
+
+	if stoargeLayout.HierarchicalNamespace != nil && stoargeLayout.GetHierarchicalNamespace().Enabled {
```

Comment:
It's better to get HierarchicalNamespace via method instead of accessing directly.

```
hierarchicalNamespace := stoargeLayout.GetHierarchicalNamespace()
if hierarchicalNamespace != nil && hierarchicalNamespace.Enabled {
...
}
```